eth/tracers/recover.go

```diff
@@ -0,0 +1,238 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package tracers
+
+import (
+	"fmt"
+	"math/big"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/tracing"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// recoverTracer is a live tracer that acts as a proxy: it delegates all hook calls
+// to a user‐defined ("child") tracer while catching any panics.
+type recoverTracer struct {
+	node  *node.Node
+	child *tracing.Hooks
+}
+
+// NewRecoverTracer instantiates a recoverTracer and returns a wrapped tracing.Hooks.
+// Only hook fields which are non-nil in the child are replaced with the corresponding
+// recoverTracer method.
+func NewRecoverTracer(node *node.Node, child *tracing.Hooks) (*tracing.Hooks, error) {
+	if child == nil {
+		return nil, fmt.Errorf("child tracer is nil")
+	}
+	rt := &recoverTracer{node: node, child: child}
+	return rt.wrapHooks()
+}
+
+// wrapHooks creates a new hooks struct with safe wrappers
+// for each non-nil hook in the child tracer.
+func (rt *recoverTracer) wrapHooks() (*tracing.Hooks, error) {
+	childVal := reflect.Indirect(reflect.ValueOf(rt.child))
+	hookType := childVal.Type()
+
+	// Create a new Hooks value.
+	newHooks := reflect.New(hookType).Elem()
+	// Get the recoverTracer's method set.
+	rtVal := reflect.ValueOf(rt)
+
+	// Iterate through each field of the Hooks struct.
+	for i := 0; i < hookType.NumField(); i++ {
+		field := hookType.Field(i)
+		childField := childVal.Field(i)
+		// If the child's hook is set, then wrap it.
+		if !childField.IsNil() {
+			// Look for the recoverTracer method with the same name.
+			methodVal := rtVal.MethodByName(field.Name)
+			// If the method exists and its type matches the hook's type, use it.
+			if methodVal.IsValid() && methodVal.Type() == field.Type {
+				fmt.Printf("Setting method %s for field %s\n", methodVal, field.Name)
+				newHooks.Field(i).Set(methodVal)
+			} else {
+				return nil, fmt.Errorf("method %s not available on recovery tracer", field.Name)
+			}
+		}
+	}
+
+	// Return the wrapped Hooks.
+	ret := newHooks.Addr().Interface().(*tracing.Hooks)
+	return ret, nil
+}
+
+// safeCall is a helper wrapping a hook call in a defer/recover block.
+// If the call panics, it logs the panic. If shutdown is true, it also initiates
+// node shutdown.
+func (rt *recoverTracer) safeCall(name string, shutdown bool, fn func()) {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Error(fmt.Sprintf("panic in child tracer during %s: %v", name, r))
+			if shutdown {
+				rt.node.Close()
+			}
```

Comment:
@fjl can you please check out this part?

Reply:
please also print stack trace!

Reply:
fixed